sunpy/map/maputils.py

```diff
@@ -1,7 +1,259 @@
+    edge_pixels = list(chain.from_iterable([edges["lhs"], edges["rhs"], edges["top"], edges["bottom"]]))
+    x = [p[0] for p in edge_pixels] * u.pix
+    y = [p[1] for p in edge_pixels] * u.pix
+
+    # Calculate the edge of the world
+    edge_of_world = smap.pixel_to_world(x, y)
+
+    # Calculate the distance of the edge of the world in solar radii
+    distance = u.R_sun * np.sqrt(edge_of_world.Tx ** 2 + edge_of_world.Ty ** 2) / smap.rsun_obs
+
+    # Test if all the edge pixels are more than one solar radius distant
+    # and that the whole map is not all off disk.
+    return np.all(distance > 1*u.R_sun) and ~is_all_off_disk(smap)
+
+
+@u.quantity_input
+def coordinate_is_on_disk(coordinate, scale: u.arcsecond):
+    """
+    Checks if the helioprojective Cartesian coordinate is on disk.
+
+    The check is performed by comparing the coordinate's distance
+    from the center of the Sun to the solar radius.
+
+    Parameters
+    ----------
+    coordinate : `~astropy.coordinates.SkyCoord`, `~sunpy.coordinates.frames.Helioprojective`
+        The input coordinate. The coordinate frame must be
+        `~sunpy.coordinates.Helioprojective`.
+
+    scale : `~astropy.units.Quantity`
+        The pixel scale size in units of arcseconds.
+
+    Returns
+    -------
+    `~bool`
+        Returns `True` if the coordinate is on disk, `False` otherwise.
+    """
+    # Calculate the radii of every pixel in helioprojective Cartesian
+    # co-ordinate distance units.
+    return u.R_sun * (np.sqrt(coordinate.Tx ** 2 + coordinate.Ty ** 2) / scale) < 1 * u.R_sun
+
+
+def is_all_off_disk(smap):
+    """
+    Checks if the entire `~sunpy.map.Map` is off the solar disk.
+
+    The check is performed by calculating the distance of every
+    pixel from the center of the Sun. If they are all off-disk,
+    then the function returns `True`. Otherwise, the function
+    returns `False`.
+
+    Parameters
+    ----------
+    smap : `~sunpy.map.GenericMap`
+        A map in helioprojective Cartesian coordinates.
+
+    Returns
+    -------
+    `~bool`
+        Returns `True` if all map pixels are strictly more than one solar radius
+        away from the center of the Sun.
+    """
+    return np.all(~coordinate_is_on_disk(all_coordinates_from_map(smap), smap.rsun_obs))
+
+
+def is_all_on_disk(smap):
+    """
+    Checks if the entire map is on the solar disk.
+
+    The check is performed by calculating the distance of every pixel
+    from the center of the Sun. If they are all on-disk, then the
+    function returns `True`. Otherwise, the function returns `False`.
+
+    Parameters
+    ----------
+    smap : `~sunpy.map.GenericMap`
+        A map in helioprojective Cartesian coordinates.
+
+    Returns
+    -------
+    `~bool`
+        Returns `True` if all map pixels are strictly less than one solar radius
+        away from the center of the Sun.
+    """
+    return np.all(coordinate_is_on_disk(all_coordinates_from_map(smap), smap.rsun_obs))
```

Comment:
Rsun_obs is not a scale. So this doesn't seem like it should work.

Reply:
Also should we be trusting the maps rsun_obs?

Reply:
Rsun_obs is the radius of the Sun as measured in image pixels and is typically provided in the FITS header.  Trust is a separate issue.

Reply:
The choice of the word 'scale' is not good.  When I was originally coding this I wanted `coordinate_is_on_disk` to be useful also for coronagraphic images.  In that scenario a use case might be to know if the pixel is sitting on the occulting disk.  Therefore you don't want to compare a pixel's radial distance from Sun center to the radius of the Sun in arcseconds; you want to use the radius of the occulting disk.  So the distance measure you are comparing against would change, and so I chose the word 'scale'.  So I could change the name of this function to make this function more generic-sounding, or just change the word 'scale' to 'rsun_obs'.

Reply:
Speaking of coronagraph images, perhaps we could mask out the occulting disk. That might be a useful feature to provide. I've created an issue to discuss #3098.

Reply:
See my comment above. I think all of these functions should use the observer position and not the plate scale which does include that information but also confusingly includes the pixel size.

Reply:
I've changed the function completely so that rsun_obs is not required.